pkg/odo/genericclioptions/runnable.go

```diff
@@ -26,9 +44,28 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) {
 	// LogErrorAndExit is used so that we get -o (jsonoutput) for cmds which have json output implemented
 	util.LogErrorAndExit(checkConflictingFlags(cmd), "")
 	// Run completion, validation and run.
-	util.LogErrorAndExit(o.Complete(cmd.Name(), cmd, args), "")
-	util.LogErrorAndExit(o.Validate(), "")
-	util.LogErrorAndExit(o.Run(), "")
+	// Only upload data to segment for completion and validation if a non-nil error is returned.
+	err = o.Complete(cmd.Name(), cmd, args)
+	if err != nil {
+		uploadToSegmentAndLog(cmd, err, startTime)
+	}
+	err = o.Validate()
+	if err != nil {
+		uploadToSegmentAndLog(cmd, err, startTime)
+	}
+	uploadToSegmentAndLog(cmd, o.Run(), startTime)
+}
+
+// uploadToSegmentAndLog uploads the data to segment
+func uploadToSegmentAndLog(cmd *cobra.Command, err error, startTime time.Time) {
+	if serr := segmentClient.Upload(cmd.CommandPath(), time.Since(startTime), err); serr != nil {
+		klog.Errorf("Cannot send data to telemetry: %v", serr)
+	}
+	// If the error is not nil, client will be closed so that data can be sent before the program exits.
+	if err != nil {
+		segmentClient.Close()
```

Comment:
I saw a differ that closes the client already above, so do we need to close it here?

Reply:
Yes, we do. The `segmentClient.Upload` function queues the data to be sent to the segment and sends it once we call the `segmentClient.Close` and close the connection.

`uploadToSegmentAndLog` uploads the data to segment and also calls `LogErrorAndExit` which would exit the program in case err!=nil, therefore not going back to the deferred close call and not sending the data to segment. So in case err!=nil, we want to close the segment connection and send the data before exiting the program.